proctor-common/src/main/java/com/indeed/proctor/common/model/Payload.java

```diff
@@ -326,7 +326,42 @@ public boolean equals(final Object o) {
                 Objects.equals(stringValue, payload.stringValue) &&
                 Arrays.equals(stringArray, payload.stringArray) &&
                 Objects.equals(json, payload.json) &&
-                Objects.equals(map, payload.map);
+                comparePayloadMap(map, payload.map);
+    }
+
+    private static boolean comparePayloadMap(final Map<String, Object> a, final Map<String, Object> b) {
+        if (a != null && b != null && a.size() == b.size()) {
+            boolean isEqual = true;
+            for (final String keyString : a.keySet()) {
+                final Object aValue = a.get(keyString);
+                final Object bValue = b.get(keyString);
+                if (aValue instanceof Double || aValue instanceof Long || aValue instanceof String) {
+                    isEqual &= Objects.equals(aValue, bValue);
+                } else if (aValue instanceof ArrayList && bValue instanceof ArrayList) {
+                    isEqual &= Objects.equals(aValue, bValue);
+                } else if (aValue instanceof ArrayList) {
+                    isEqual &= Arrays.equals(((ArrayList<?>) aValue).toArray(), (Object[]) bValue);
+                } else if (bValue instanceof ArrayList) {
+                    isEqual &= Arrays.equals(((ArrayList<?>) bValue).toArray(), (Object[]) aValue);
+                } else if (aValue instanceof Double[]){
+                    final Double[] aArr = (Double[]) aValue;
+                    final Double[] bArr = (Double[]) bValue;
+                    isEqual &= Arrays.equals(aArr, bArr);
+                } else if (aValue instanceof Long[]){
+                    final Long[] aArr = (Long[]) aValue;
+                    final Long[] bArr = (Long[]) bValue;
+                    isEqual &= Arrays.equals(aArr, bArr);
+                } else if (aValue instanceof String[]) {
+                    final String[] aArr = (String[]) aValue;
+                    final String[] bArr = (String[]) bValue;
+                    isEqual &= Arrays.equals(aArr, bArr);
+                } else {
+                    isEqual &= Objects.equals(aValue, bValue);
+                }
+            }
+            return isEqual;
+        }
+        return Objects.equals(a, b);
```

Comment:
Should we spend more time optimizing this method?
For instances:
if(a == null || b == null || a.size() != b.size()) return false;
And for each loop, if we see a not equal, we can immediately return false. So that we don't have to loop via all keys

Reply:
> if(a == null || b == null || a.size() != b.size()) return false;
Objects.equals(a, b) will handle this
> And for each loop, if we see a not equal, we can immediately return false. So that we don't have to loop via all keys
Fixed the loop to be more efficient and break when a value is not equal

Reply:
Sounds good